Bad credentials produce a crash

```
10-26 09:00:03.604 5868-13705/org.eyeseetea.surveillance_cnm_bb D/.LoginActivity: onInvalidCredentials
10-26 09:00:03.609 5868-13705/org.eyeseetea.surveillance_cnm_bb E/CrashlyticsCore: Failed to execute task.
                                                                                   java.lang.InterruptedException
                                                                                       at java.util.concurrent.FutureTask.awaitDone(FutureTask.java:391)
                                                                                       at java.util.concurrent.FutureTask.get(FutureTask.java:163)
                                                                                       at com.crashlytics.android.core.CrashlyticsExecutorServiceWrapper.executeSyncLoggingException(CrashlyticsExecutorServiceWrapper.java:47)
                                                                                       at com.crashlytics.android.core.CrashlyticsUncaughtExceptionHandler.uncaughtException(CrashlyticsUncaughtExceptionHandler.java:235)
                                                                                       at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1068)
                                                                                       at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1063)
                                                                                       at rx.internal.schedulers.ScheduledAction.run(ScheduledAction.java:66)
                                                                                       at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
                                                                                       at java.util.concurrent.FutureTask.run(FutureTask.java:23dPoolExecutor.java:1133)
                                                                                      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
                                                                                      at java.lang.Thread.run(Thread.java:761)
                                                                                   Caused by: rx.exceptions.OnErrorFailedException: Error occurred when trying to propagate error to Observer.onError
                                                                                      at rx.observers.SafeSubscriber._onError(SafeSubscriber.java:194)
                                                                                      at rx.observers.SafeSubscriber.onError(SafeSubscriber.java:120)
                                                                                      at rx.internal.operators.OperatorObserveOn$ObserveOnSubscriber.pollQueue(OperatorObserveOn.java:191)
                                                                                      at rx.internal.operators.OperatorObserveOn$ObserveOnSubscriber$2.call(OperatorObserveOn.java:162)
                                                                                      at rx.internal.schedulers.ScheduledAction.run(ScheduledAction.java:55)
                                                                                      at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428) 
                                                                                      at java.util.concurrent.FutureTask.run(FutureTask.java:237) 
                                                                                      at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:272) 
                                                                                      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133) 
                                                                                      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607) 
                                                                                      at java.lang.Thread.run(Thread.java:761) 
                                                                                   Caused by: rx.exceptions.CompositeException: 2 exceptions occurred. 
                                                                                      at rx.observers.SafeSubscriber._onError(SafeSubscriber.java:194) 
                                                                                      at rx.observers.SafeSubscriber.onError(SafeSubscriber.java:120) 
                                                                                      at rx.internal.operators.OperatorObserveOn$ObserveOnSubscriber.pollQueue(OperatorObserveOn.java:191) 
                                                                                      at rx.internal.operators.OperatorObserveOn$ObserveOnSubscriber$2.call(OperatorObserveOn.java:162) 
                                                                                      at rx.internal.schedulers.ScheduledAction.run(ScheduledAction.java:55) 
                                                                                      at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428) 
                                                                                      at java.util.concurrent.FutureTask.run(FutureTask.java:237) 
                                                                                      at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:272) 
                                                                                      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133) 
                                                                                      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607) 
                                                                                      at java.lang.Thread.run(Thread.java:761) 
                                                                                   Caused by: rx.exceptions.CompositeException$CompositeExceptionCausalChain: Chain of Causes for CompositeException In Order Received =>
                                                                                      at android.util.Log.printlns(Log.java:415)
                                                                                      at com.android.internal.os.RuntimeInit.Clog_e(RuntimeInit.java:61)
                                                                                      at com.android.internal.os.RuntimeInit.-wrap0(RuntimeInit.java)
                                                                                      at com.android.internal.os.RuntimeInit$UncaughtHandler.uncaughtException(RuntimeInit.java:86)
                                                                                      at com.crashlytics.android.core.CrashlyticsUncaughtExceptionHandler.uncaughtException(CrashlyticsUncaughtExceptionHandler.java:249)
                                                                                      at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1068)
                                                                                      at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:1063)
                                                                                      at rx.internal.schedulers.ScheduledAction.run(ScheduledAction.java:66)
                                                                                      at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428) 
                                                                                      at java.util.concurrent.FutureTask.run(FutureTask.java:237) 
                                                                                      at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:272) 
                                                                                      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133) 
                                                                                      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607) 
                                                                                      at java.lang.Thread.run(Thread.java:761) 
                                                                                   Caused by: org.hisp.dhis.client.sdk.core.common.network.ApiException: 401 Unauthorized
                                                                                      at org.hisp.dhis.client.sdk.android.api.network.NetworkUtils.call(NetworkUtils.java:179)
                                                                                      at org.hisp.dhis.client.sdk.android.user.UserAccountApiClientImpl.getUserAccount(UserAccountApiClientImpl.java:54)
                                                                                      at org.hisp.dhis.client.sdk.core.user.UserAccountControllerImpl.pull(UserAccountControllerImpl.java:73)
                                                                                      at org.hisp.dhis.client.sdk.android.user.CurrentUserInteractorImpl$1.call(CurrentUserInteractorImpl.java:94)
                                                                                      at org.hisp.dhis.client.sdk.android.user.CurrentUserInteractorImpl$1.call(CurrentUserInteractorImpl.java:84)
                                                                                      at org.hisp.dhis.client.sdk.android.api.utils.DefaultOnSubscribe.call(DefaultOnSubscribe.java:39)
                                                                                      at org.hisp.dhis.client.sdk.android.api.utils.DefaultOnSubscribe.call(DefaultOnSubscribe.java:34)
                                                                                      at rx.Observable.unsafeSubscribe(Observable.java:8098)
                                                                                      at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:232)
                                                                                      at rx.internal.operators.OperatorMerge$MergeSubscriber.onNext(OperatorMerge.java:142)
                                                                                      at rx.internal.operators.OperatorMap$1.onNext(OperatorMap.java:54)
                                                                                      at org.hisp.dhis.client.sdk.android.api.utils.DefaultOnSubscribe.call(DefaultOnSubscribe.java:39)
                                                                                      at org.hisp.dhis.client.sdk.android.api.utils.DefaultOnSubscribe.call(DefaultOnSubscribe.java:34)
                                                                                      at rx.Observable$2.call(Observable.java:162)
                                                                                      at rx.Observable$2.call(Observable.java:154)
                                                                                      at rx.Observable$2.call(Observable.java:162)
                                                                                      at rx.Observable$2.call(Observable.java:154)
                                                                                      at rx.Observable.unsafeSubscribe(Observable.java:8098)
                                                                                      at rx.internal.operators.OperatorSubscribeOn$1$1.call(OperatorSubscribeOn.java:62)
                                                                                      at rx.internal.schedulers.ScheduledAction.run(ScheduledAction.java:55)
                                                                                      at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428) 
                                                                                      at java.util.concurrent.FutureTask.run(FutureTask.java:237) 
                                                                                      at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:272) 
                                                                                      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133) 
                                                                                      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607) 
                                                                                      at java.lang.Thread.run(Thread.java:761) 

```

Caused by: org.hisp.dhis.client.sdk.core.common.network.ApiException: 401 Unauthorized
         at org.hisp.dhis.client.sdk.android.api.network.NetworkUtils.call(NetworkUtils.java:179)
         at org.hisp.dhis.client.sdk.android.user.UserAccountApiClientImpl.getUserAccount(UserAccountApiClientImpl.java:54)
         at org.hisp.dhis.client.sdk.core.user.UserAccountControllerImpl.pull(UserAccountControllerImpl.java:73)

--- a/app/src/cnm/java/org/eyeseetea/malariacare/domain/usecase/LoginUseCase.java
+++ b/app/src/cnm/java/org/eyeseetea/malariacare/domain/usecase/LoginUseCase.java
@@ -34,6 +34,31 @@ public class LoginUseCase extends ALoginUseCase implements UseCase {
         mAsyncExecutor.run(this);
     }
 
+    @Override
+    public void run() {
+        if (mCredentials == null) {
+            onErrorCallback(mCallback,
+                    new IllegalArgumentException("Credentials could not be null"));
+        } else {
+            mAuthenticationManager.login(mCredentials,
+                    new IAuthenticationManager.Callback<UserAccount>() {
+                        @Override
+                        public void onSuccess(UserAccount userAccount) {
+                            if (!mCredentials.isDemoCredentials()) {
+                                logoutAndHardcodedLogin(mCredentials, mCallback);
+                            } else {
+                                mCallback.onLoginSuccess();
+                            }
+                        }
+
+                        @Override
+                        public void onError(Throwable throwable) {
+                            onErrorCallback(mCallback, throwable);
+                        }
+                    });
+        }
+    }
+
     private void logoutAndHardcodedLogin(final Credentials credentials,
             final Callback loginCallback) {
         mAuthenticationManager.logout(new IAuthenticationManager.Callback<Void>() {
@@ -127,28 +152,5 @@ public class LoginUseCase extends ALoginUseCase implements UseCase {
             }
         });
     }
-    @Override
-    public void run() {
-        if (mCredentials == null) {
-            onErrorCallback(mCallback,
-                    new IllegalArgumentException("Credentials could not be null"));
-        } else {
-            mAuthenticationManager.login(mCredentials,
-                    new IAuthenticationManager.Callback<UserAccount>() {
-                        @Override
-                        public void onSuccess(UserAccount userAccount) {
-                            if (!mCredentials.isDemoCredentials()) {
-                                logoutAndHardcodedLogin(mCredentials, mCallback);
-                            } else {
-                                mCallback.onLoginSuccess();
-                            }
-                        }
 
-                        @Override
-                        public void onError(Throwable throwable) {
-                            onErrorCallback(mCallback, throwable);
-                        }
-                    });
-        }
-    }
 }

--- a/app/src/cnm/java/org/eyeseetea/malariacare/domain/usecase/LoginUseCase.java
+++ b/app/src/cnm/java/org/eyeseetea/malariacare/domain/usecase/LoginUseCase.java
@@ -1,6 +1,8 @@
 package org.eyeseetea.malariacare.domain.usecase;
 
 import org.eyeseetea.malariacare.domain.boundary.IAuthenticationManager;
+import org.eyeseetea.malariacare.domain.boundary.executors.IAsyncExecutor;
+import org.eyeseetea.malariacare.domain.boundary.executors.IMainExecutor;
 import org.eyeseetea.malariacare.domain.entity.Credentials;
 import org.eyeseetea.malariacare.domain.entity.UserAccount;
 import org.eyeseetea.malariacare.domain.exception.ConfigJsonIOException;
@@ -11,36 +13,25 @@ import java.net.MalformedURLException;
 import java.net.UnknownHostException;
 
 
-public class LoginUseCase extends ALoginUseCase {
+public class LoginUseCase extends ALoginUseCase implements UseCase {
     private IAuthenticationManager mAuthenticationManager;
+    private IAsyncExecutor mAsyncExecutor;
+    private IMainExecutor mMainExecutor;
+    Credentials mCredentials;
+    Callback mCallback;
 
-    public LoginUseCase(IAuthenticationManager authenticationManager) {
+    public LoginUseCase(IAuthenticationManager authenticationManager, IAsyncExecutor asyncExecutor,
+            IMainExecutor mainExecutor) {
+        mMainExecutor = mainExecutor;
+        mAsyncExecutor = asyncExecutor;
         mAuthenticationManager = authenticationManager;
     }
 
     @Override
     public void execute(final Credentials credentials, final Callback loginCallback) {
-        if (credentials == null) {
-            onErrorCallback(loginCallback,
-                    new IllegalArgumentException("Credentials could not be null"));
-        } else {
-            mAuthenticationManager.login(credentials,
-                    new IAuthenticationManager.Callback<UserAccount>() {
-                        @Override
-                        public void onSuccess(UserAccount userAccount) {
-                            if (!credentials.isDemoCredentials()) {
-                                logoutAndHardcodedLogin(credentials, loginCallback);
-                            } else {
-                                loginCallback.onLoginSuccess();
-                            }
-                        }
-
-                        @Override
-                        public void onError(Throwable throwable) {
-                            onErrorCallback(loginCallback, throwable);
-                        }
-                    });
-        }
+        mCredentials = credentials;
+        mCallback = loginCallback;
+        mAsyncExecutor.run(this);
     }
 
     private void logoutAndHardcodedLogin(final Credentials credentials,
@@ -76,19 +67,88 @@ public class LoginUseCase extends ALoginUseCase {
     }
 
     private void onErrorCallback(final Callback callback, Throwable throwable) {
+        throwable.printStackTrace();
         if (throwable instanceof MalformedURLException
                 || throwable instanceof UnknownHostException) {
-            callback.onServerURLNotValid();
+            onServerURLNotValid();
         } else if (throwable instanceof InvalidCredentialsException) {
-            callback.onInvalidCredentials();
+            onInvalidCredentials();
         } else if (throwable instanceof NetworkException) {
-            callback.onNetworkError();
+            onNetworkError();
         } else if (throwable instanceof ConfigJsonIOException) {
-            callback.onConfigJsonInvalid();
+            onConfigJsonInvalid();
         } else {
-            throwable.printStackTrace();
-            callback.onUnexpectedError();
+            onUnexpectedError();
         }
 
+    }
+
+    private void onUnexpectedError() {
+        mMainExecutor.run(new Runnable() {
+            @Override
+            public void run() {
+                mCallback.onUnexpectedError();
+            }
+        });
+    }
+
+    private void onConfigJsonInvalid() {
+        mMainExecutor.run(new Runnable() {
+            @Override
+            public void run() {
+                mCallback.onConfigJsonInvalid();
+            }
+        });
+    }
+
+    private void onNetworkError() {
+        mMainExecutor.run(new Runnable() {
+            @Override
+            public void run() {
+                mCallback.onNetworkError();
+            }
+        });
+    }
+
+    private void onInvalidCredentials() {
+        mMainExecutor.run(new Runnable() {
+            @Override
+            public void run() {
+                mCallback.onInvalidCredentials();
+            }
+        });
+    }
+
+    private void onServerURLNotValid() {
+        mMainExecutor.run(new Runnable() {
+            @Override
+            public void run() {
+                mCallback.onServerURLNotValid();
+            }
+        });
+    }
+    @Override
+    public void run() {
+        if (mCredentials == null) {
+            onErrorCallback(mCallback,
+                    new IllegalArgumentException("Credentials could not be null"));
+        } else {
+            mAuthenticationManager.login(mCredentials,
+                    new IAuthenticationManager.Callback<UserAccount>() {
+                        @Override
+                        public void onSuccess(UserAccount userAccount) {
+                            if (!mCredentials.isDemoCredentials()) {
+                                logoutAndHardcodedLogin(mCredentials, mCallback);
+                            } else {
+                                mCallback.onLoginSuccess();
+                            }
+                        }
+
+                        @Override
+                        public void onError(Throwable throwable) {
+                            onErrorCallback(mCallback, throwable);
+                        }
+                    });
+        }
     }
 }

--- a/app/src/cnm/java/org/eyeseetea/malariacare/strategies/BaseActivityStrategy.java
+++ b/app/src/cnm/java/org/eyeseetea/malariacare/strategies/BaseActivityStrategy.java
@@ -10,10 +10,14 @@ import org.eyeseetea.malariacare.R;
 import org.eyeseetea.malariacare.data.authentication.AuthenticationManager;
 import org.eyeseetea.malariacare.data.database.utils.PreferencesState;
 import org.eyeseetea.malariacare.domain.boundary.IAuthenticationManager;
+import org.eyeseetea.malariacare.domain.boundary.executors.IAsyncExecutor;
+import org.eyeseetea.malariacare.domain.boundary.executors.IMainExecutor;
 import org.eyeseetea.malariacare.domain.entity.Credentials;
 import org.eyeseetea.malariacare.domain.usecase.LoginUseCase;
 import org.eyeseetea.malariacare.domain.usecase.LogoutUseCase;
 import org.eyeseetea.malariacare.layout.utils.LayoutUtils;
+import org.eyeseetea.malariacare.presentation.executors.AsyncExecutor;
+import org.eyeseetea.malariacare.presentation.executors.UIThreadExecutor;
 import org.eyeseetea.malariacare.services.SurveyService;
 
 public class BaseActivityStrategy extends ABaseActivityStrategy {
@@ -42,7 +46,10 @@ public class BaseActivityStrategy extends ABaseActivityStrategy {
 
         mAuthenticationManager = new AuthenticationManager(mBaseActivity);
         mLogoutUseCase = new LogoutUseCase(mAuthenticationManager);
-        mLoginUseCase = new LoginUseCase(mAuthenticationManager);
+        IAsyncExecutor asyncExecutor = new AsyncExecutor();
+        IMainExecutor mainExecutor = new UIThreadExecutor();
+
+        mLoginUseCase = new LoginUseCase(mAuthenticationManager, asyncExecutor, mainExecutor);
     }
 
     @Override

--- a/app/src/cnm/java/org/eyeseetea/malariacare/strategies/LoginActivityStrategy.java
+++ b/app/src/cnm/java/org/eyeseetea/malariacare/strategies/LoginActivityStrategy.java
@@ -7,9 +7,13 @@ import org.eyeseetea.malariacare.LoginActivity;
 import org.eyeseetea.malariacare.ProgressActivity;
 import org.eyeseetea.malariacare.SettingsActivity;
 import org.eyeseetea.malariacare.domain.boundary.IAuthenticationManager;
+import org.eyeseetea.malariacare.domain.boundary.executors.IAsyncExecutor;
+import org.eyeseetea.malariacare.domain.boundary.executors.IMainExecutor;
 import org.eyeseetea.malariacare.domain.entity.Credentials;
 import org.eyeseetea.malariacare.domain.usecase.LoadUserAndCredentialsUseCase;
 import org.eyeseetea.malariacare.domain.usecase.LoginUseCase;
+import org.eyeseetea.malariacare.presentation.executors.AsyncExecutor;
+import org.eyeseetea.malariacare.presentation.executors.UIThreadExecutor;
 
 public class LoginActivityStrategy extends ALoginActivityStrategy {
     public LoginActivityStrategy(LoginActivity loginActivity) {
@@ -68,7 +72,10 @@ public class LoginActivityStrategy extends ALoginActivityStrategy {
 
     @Override
     public void initLoginUseCase(IAuthenticationManager authenticationManager) {
-        loginActivity.mLoginUseCase = new LoginUseCase(authenticationManager);
+        IAsyncExecutor asyncExecutor = new AsyncExecutor();
+        IMainExecutor mainExecutor = new UIThreadExecutor();
+
+        loginActivity.mLoginUseCase = new LoginUseCase(authenticationManager, asyncExecutor, mainExecutor);
     }
 
 

